relay-server/src/services/proxy_processor.rs

```diff
@@ -0,0 +1,163 @@
+use std::error::Error;
+use std::sync::Arc;
+
+use relay_config::Config;
+use relay_statsd::metric;
+use relay_system::{Addr, Service};
+
+use crate::envelope::{ContentType, Envelope, EnvelopeError, Item, ItemType};
+use crate::extractors::{PartialDsn, RequestMeta};
+use crate::managed::ManagedEnvelope;
+use crate::services::outcome::{DiscardReason, Outcome, TrackOutcome};
+use crate::services::processor::{
+    EnvelopeProcessor, ProcessEnvelope, ProcessingGroup, SendEnvelope, SubmitClientReports,
+    encode_payload,
+};
+
+use crate::services::projects::cache::ProjectCacheHandle;
+use crate::services::upstream::{SendRequest, UpstreamRelay};
+use crate::statsd::RelayTimers;
+
+/// Service implementing the [`EnvelopeProcessor`] interface.
+///
+/// Analog to [`crate::services::processor::EnvelopeProcessorService`] this service handles messages when Relay is run in
+/// proxy mode.
+pub struct ProxyProcessorService {
+    inner: InnerProxyProcessor,
+}
+
+struct InnerProxyProcessor {
+    config: Arc<Config>,
+    project_cache: ProjectCacheHandle,
+    addrs: ProxyAddrs,
+}
+
+/// Contains the addresses of services that the proxy-processor publishes to.
+pub struct ProxyAddrs {
+    pub outcome_aggregator: Addr<TrackOutcome>,
+    pub upstream_relay: Addr<UpstreamRelay>,
+}
+
+impl Default for ProxyAddrs {
+    fn default() -> Self {
+        ProxyAddrs {
+            outcome_aggregator: Addr::dummy(),
+            upstream_relay: Addr::dummy(),
+        }
+    }
+}
+
+impl ProxyProcessorService {
+    /// Creates a multi-threaded proxy processor.
+    pub fn new(config: Arc<Config>, project_cache: ProjectCacheHandle, addrs: ProxyAddrs) -> Self {
+        Self {
+            inner: InnerProxyProcessor {
+                project_cache,
+                addrs,
+                config,
+            },
+        }
+    }
+
+    fn handle_process_envelope(&self, message: ProcessEnvelope) {
+        let wait_time = message.envelope.age();
+        metric!(timer(RelayTimers::EnvelopeWaitTime) = wait_time);
+
+        let scoping = message.envelope.scoping();
+        for (_, envelope) in ProcessingGroup::split_envelope(
```

Comment:
Why does proxy even need to split an envelope? Couldn't it just forward it unaltered?